### PR TITLE
Add parsed url to PubGrubPackage

### PR DESCRIPTION
**Pull request by konstin**
Avoid reparsing urls by storing the parsed parts across resolution on `PubGrubPackage`.

Part 1 of #3408

### CODE REVIEW

File: crates/uv-resolver/src/resolver/urls.rs [hunk 1]
```diff
@@ -41,27 +56,77 @@ impl Urls {
...
+                    let url = VerbatimParsedUrl {
+                        parsed_url: ParsedUrl::Archive(ParsedArchiveUrl {
+                            url: location.clone(),
+                            subdirectory: subdirectory.clone(),
+                        }),
+                        verbatim: url.clone(),
+                    };
                     if let Some(previous) = urls.insert(requirement.name.clone(), url.clone()) {
-                        if !is_equal(&previous, url) {
+                        // TODO(merge): use proper check
```
Comment (charliermarsh) [inline crates/uv-resolver/src/resolver/urls.rs:72]: What's this?

File: crates/distribution-types/src/parsed_url.rs [hunk 2]
```diff
@@ -22,13 +24,21 @@ pub enum ParsedUrlError {
     UrlParse(String, #[source] url::ParseError),
 }
 
+#[derive(Debug, Clone, Derivative)]
+#[derivative(Hash, PartialEq, Eq)]
+pub struct VerbatimParsedUrl {
+    #[derivative(PartialEq = "ignore", Hash = "ignore")]
```
Comment (charliermarsh) [inline crates/distribution-types/src/parsed_url.rs:30]: Why is this necessary? Where are we relying on this? I find that this pattern tends to cause a lot of subtle bugs.

File: crates/uv-resolver/src/resolver/urls.rs (hunk 1 above)
Comment (konstin) [reply crates/uv-resolver/src/resolver/urls.rs]: At some future point in time, we should use an equality check by parts

File: crates/distribution-types/src/parsed_url.rs (hunk 2 above)
Comment (konstin) [reply crates/distribution-types/src/parsed_url.rs:30]: We were previously relying on equality between two `VerbatimUrl`s. This derive preserve this behavior, so we don't get regression when the `ParsedUrl` is slightly different from the `VerbatimUrl`. Eventually, we should remove that check

Comment (charliermarsh) [reply crates/distribution-types/src/parsed_url.rs:30]: I think my question is: where are we relying on this? Any way we can make this explicit at those sites?

File: crates/uv-resolver/src/resolver/urls.rs [hunk 3]
```diff
@@ -1,36 +1,51 @@
...
-                        debug!("Allowing {} as a variant of {previous}", editable.url);
+            let editable_url = VerbatimParsedUrl {
+                parsed_url: ParsedUrl::LocalFile(ParsedLocalFileUrl {
+                    url: editable.url.to_url(),
+                    path: editable.path.clone(),
+                    editable: true,
+                }),
+                verbatim: editable.url.clone(),
+            };
+            if let Some(previous) = urls.insert(metadata.name.clone(), editable_url.clone()) {
```
Comment (charliermarsh) [inline crates/uv-resolver/src/resolver/urls.rs:37]: The construction of `VerbatimParsedUrl` seems awkward here, just a lot of conversion. Will it be removed eventually?

File: crates/uv-resolver/src/redirect.rs [hunk 4]
```diff
@@ -36,9 +39,50 @@ pub(crate) fn apply_redirect(url: &VerbatimUrl, redirect: Url) -> VerbatimUrl {
...
+        .map_or(GitReference::DefaultBranch, |rev| {
+            GitReference::from_rev(rev)
+        });
+    let git_url = GitUrl::new(git_url.repository().clone(), lowered_git_ref);
+
+    let Some(new_git_url) = git_url_to_precise(git_url.clone()) else {
+        if matches!(git_url.reference(), GitReference::FullCommit(_)) {
+            return url;
+        }
+        panic!("Unseen git url: {}, {:?}", url.verbatim, git_url);
```
Comment (charliermarsh) [inline crates/uv-resolver/src/redirect.rs:64]: Should we just return the original URL here instead?

File: crates/uv-resolver/src/resolver/urls.rs (hunk 3 above)
Comment (konstin) [reply crates/uv-resolver/src/resolver/urls.rs:37]: Yes, `LocalEditable` will switch away from `url: VerbatimUrl` to a `VerbatimParsedUrl`.

File: crates/uv-resolver/src/redirect.rs (hunk 4 above)
Comment (konstin) [reply crates/uv-resolver/src/redirect.rs:64]: Made this a debug assert, this would be a bug in uv but not one we have to concern the user with (we only fail the already-install checked if tha happens).

**Review by charliermarsh — APPROVED** [review]
